File: app/src/main/java/com/wkz/viewpager/transformer/TabletTransformer.java
package com.wkz.viewpager.transformer;

import android.graphics.Camera;
import android.graphics.Matrix;
import android.view.View;

import com.nineoldandroids.view.ViewHelper;

/**
 * @desc 平板的转换效果
 * @date 2017-06-30
 */
public class TabletTransformer extends AbstractBaseTransformer {

    private Matrix mMatrix = new Matrix();
    private Camera mCamera = new Camera();
    private float[] mTempFloat2 = new float[2];

    @Override
    protected void onTransform(View page, float position) {

        float rotation = (position < 0 ? 30F : -30F) * Math.abs(position);

        ViewHelper.setTranslationX(page, getOffsetXForRotation(rotation, page.getWidth(), page.getHeight()));
        ViewHelper.setPivotX(page, page.getWidth() * 0.5F);
        ViewHelper.setPivotY(page, page.getWidth() * 0.5F);
        ViewHelper.setRotationY(page, rotation);

    }

    private float getOffsetXForRotation(float degrees, int width, int height) {
        mMatrix.reset();
        mCamera.save();
        mCamera.rotateY(Math.abs(degrees));
        mCamera.getMatrix(mMatrix);
        mCamera.restore();

        mMatrix.preTranslate(-width * 0.5F, -height * 0.5F);
        mMatrix.postTranslate(width * 0.5F, height * 0.5F);
        mTempFloat2[0] = width;
        mTempFloat2[1] = height;
        mMatrix.mapPoints(mTempFloat2);
        return (width - mTempFloat2[0]) * (degrees > 0.0F ? 1.0F : -1.0F);
    }
}

File: view-pager-transformer/src/main/java/com/fphoenixcorneae/transformer/TabletTransformer.kt
package com.fphoenixcorneae.transformer

import android.graphics.Camera
import android.graphics.Matrix
import android.view.View
import com.nineoldandroids.view.ViewHelper
import kotlin.math.abs

/**
 * @desc 平板的转换效果
 * @date 2017-06-30
 */
class TabletTransformer : AbstractBaseTransformer() {
    private val mMatrix = Matrix()
    private val mCamera = Camera()
    private val mTempFloat2 = FloatArray(2)
    override fun onTransform(page: View, position: Float) {
        val rotation = (if (position < 0) 30f else -30f) * abs(position)
        ViewHelper.setTranslationX(page, getOffsetXForRotation(rotation, page.width, page.height))
        ViewHelper.setPivotX(page, page.width * 0.5f)
        ViewHelper.setPivotY(page, page.width * 0.5f)
        ViewHelper.setRotationY(page, rotation)
    }

    private fun getOffsetXForRotation(degrees: Float, width: Int, height: Int): Float {
        mMatrix.reset()
        mCamera.save()
        mCamera.rotateY(abs(degrees))
        mCamera.getMatrix(mMatrix)
        mCamera.restore()
        mMatrix.preTranslate(-width * 0.5f, -height * 0.5f)
        mMatrix.postTranslate(width * 0.5f, height * 0.5f)
        mTempFloat2[0] = width.toFloat()
        mTempFloat2[1] = height.toFloat()
        mMatrix.mapPoints(mTempFloat2)
        return (width - mTempFloat2[0]) * if (degrees > 0.0f) 1.0f else -1.0f
    }
}